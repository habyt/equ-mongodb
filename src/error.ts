import { EQUError } from "@equ-lang/parser"

export class MongoConversionError extends EQUError {
    constructor(msg?: string) {
        super(msg)
    }
}
