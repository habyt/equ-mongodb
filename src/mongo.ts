import { FilterQuery } from "mongodb"
import {
    ParseItem,
    FilterOperand,
    FilterOperator,
    ExpressionOperator,
    ExpressionItem,
    parse
} from "@equ-lang/parser"

export class MongoConversionError extends Error {
    constructor(msg?: string) {
        super(msg)
    }
}

function mongoError(msg?: string): never {
    throw new MongoConversionError(msg)
}

function convertOperator<TSchema>(
    item: FilterOperator,
    input: Array<ParseItem>
): FilterQuery<TSchema> {
    const op1 = convertOperand(input.pop() ?? mongoError(), input)
    const op2 = convertOperand(input.pop() ?? mongoError(), input)

    if (item.operator === "and") {
        return { $and: [op1, op2] } as FilterQuery<TSchema>
    }

    if (item.operator === "or") {
        return { $or: [op1, op2] } as FilterQuery<TSchema>
    }

    mongoError("unsupported operator " + item.operator)
}

function convertOperand<TSchema>(
    item: ParseItem,
    input: Array<ParseItem>
): FilterQuery<TSchema> {
    if (item.type === "operator") {
        return convertOperator(item, input)
    }

    return convertExpressions(item)
}

function convertExpressions<TSchema>(
    item: FilterOperand
): FilterQuery<TSchema> {
    const { path, expressions } = item

    const start = expressions.pop() ?? mongoError("expected expression")

    return convertExpressionOperand(start, path, expressions)
}

function convertExpressionOperator<TSchema>(
    operator: ExpressionOperator,
    path: string,
    expressions: Array<ExpressionItem>
): FilterQuery<TSchema> {
    const op1 = convertExpressionOperand(
        expressions.pop() ?? mongoError("expected expression"),
        path,
        expressions
    )
    const op2 = convertExpressionOperand(
        expressions.pop() ?? mongoError("expected expression"),
        path,
        expressions
    )

    if (operator.operator === "and") {
        return { $and: [op1, op2] } as FilterQuery<TSchema>
    }

    if (operator.operator === "or") {
        return { $or: [op1, op2] } as FilterQuery<TSchema>
    }

    mongoError("unsupported expression operator " + operator.operator)
}

function convertExpressionOperand<TSchema>(
    expression: ExpressionItem,
    path: string,
    expressions: Array<ExpressionItem>
): FilterQuery<TSchema> {
    if (expression.type === "expressionOperator") {
        return convertExpressionOperator(expression, path, expressions)
    }

    switch (expression.expressionType) {
        case "eq": {
            return { [path]: { $eq: expression.value } } as FilterQuery<TSchema>
        }
        case "gt": {
            return { [path]: { $gt: expression.value } } as FilterQuery<TSchema>
        }
        case "gte": {
            return { [path]: { $gte: expression.value } } as FilterQuery<
                TSchema
            >
        }
        case "lt": {
            return { [path]: { $lt: expression.value } } as FilterQuery<TSchema>
        }
        case "lte": {
            return { [path]: { $lte: expression.value } } as FilterQuery<
                TSchema
            >
        }
        case "ct": {
            if (expression.valueType === "number") {
                mongoError(
                    "invalid expression: " +
                        expression.expressionType +
                        " " +
                        expression.value +
                        ". Value is a number but must be string"
                )
            }
            return { [path]: { $regex: expression.value } } as FilterQuery<
                TSchema
            >
        }
        case "rgx": {
            return { [path]: { $regex: expression.value } } as FilterQuery<
                TSchema
            >
        }
        case "neq": {
            return {
                [path]: { $not: { $eq: expression.value } }
            } as FilterQuery<TSchema>
        }
        case "ex": {
            return { [path]: { $exists: expression.value } } as FilterQuery<
                TSchema
            >
        }
        default: {
            mongoError("cannot convert " + expression.expressionType)
        }
    }
}

export function equToMongoDB<TSchema>(equ: string): FilterQuery<TSchema> {
    const result = parse(equ)

    return toMongo(result)
}

function toMongo<TSchema>(input: Array<ParseItem>): FilterQuery<TSchema> {
    const start = input.pop() ?? mongoError("expected something")

    return convertOperand(start, input)
}
