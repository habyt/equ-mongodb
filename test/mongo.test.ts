import { equToMongoDB } from "../src/mongo"

const example = `name.first[eq:"foo"|eq:"bar"]|email[ct:"foo",ct:"bar"],age[(gt:4.5,lt:-10)|eq:15]`

describe("equ-mongodb", () => {
    it("should translate the example to a mongodb query", () => {
        const result = equToMongoDB(example)
        expect(result).toMatchSnapshot()
    })

    it("should translate a simple query", () => {
        const result = equToMongoDB("path[eq:1]")
        expect(result).toMatchSnapshot()
    })

    it("should translate a simple gte query", () => {
        const result = equToMongoDB("path[gte:1]")
        expect(result).toMatchSnapshot()
    })

    it("should translate a simple lte query", () => {
        const result = equToMongoDB("path[lte:1]")
        expect(result).toMatchSnapshot()
    })

    it("should throw an error on using contains with a number", () => {
        expect(() => equToMongoDB("path[ct:1]")).toThrowErrorMatchingSnapshot()
    })

    it("should convert a regex expression", () => {
        const result = equToMongoDB("path[rgx:\"as.f\"]")
        expect(result).toMatchSnapshot()
    })

    it("should convert a neq expression", () => {
        const result = equToMongoDB("path[neq:1]")
        expect(result).toMatchSnapshot()
    })

    it("should convert an ex expression", () => {
        const result = equToMongoDB("path[ex:true]")
        console.log(result)
        expect(result).toMatchSnapshot()

        const result2 = equToMongoDB("path[ex:false]")
        console.log(result2)
        expect(result).toMatchSnapshot()
    })
})

