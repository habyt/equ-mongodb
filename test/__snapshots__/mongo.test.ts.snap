// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`equ-mongodb should convert a neq expression 1`] = `
Object {
  "path": Object {
    "$ne": 1,
  },
}
`;

exports[`equ-mongodb should convert a regex expression 1`] = `
Object {
  "path": Object {
    "$regex": "as.f",
  },
}
`;

exports[`equ-mongodb should convert an ex expression 1`] = `
Object {
  "path": Object {
    "$exists": true,
  },
}
`;

exports[`equ-mongodb should convert an ex expression 2`] = `
Object {
  "path": Object {
    "$exists": true,
  },
}
`;

exports[`equ-mongodb should translate a simple gte query 1`] = `
Object {
  "path": Object {
    "$gte": 1,
  },
}
`;

exports[`equ-mongodb should translate a simple lte query 1`] = `
Object {
  "path": Object {
    "$lte": 1,
  },
}
`;

exports[`equ-mongodb should translate a simple query 1`] = `
Object {
  "path": Object {
    "$eq": 1,
  },
}
`;

exports[`equ-mongodb should translate the example to a mongodb query 1`] = `
Object {
  "$and": Array [
    Object {
      "$or": Array [
        Object {
          "age": Object {
            "$eq": 15,
          },
        },
        Object {
          "$and": Array [
            Object {
              "age": Object {
                "$lt": -10,
              },
            },
            Object {
              "age": Object {
                "$gt": 4.5,
              },
            },
          ],
        },
      ],
    },
    Object {
      "$or": Array [
        Object {
          "$and": Array [
            Object {
              "email": Object {
                "$regex": "bar",
              },
            },
            Object {
              "email": Object {
                "$regex": "foo",
              },
            },
          ],
        },
        Object {
          "$or": Array [
            Object {
              "name.first": Object {
                "$eq": "bar",
              },
            },
            Object {
              "name.first": Object {
                "$eq": "foo",
              },
            },
          ],
        },
      ],
    },
  ],
}
`;
